server/src/inmantals/server.py

```diff
@@ -130,68 +312,29 @@ def flatten(self, line, char):
-
-        modulepath = ["libs", os.path.dirname(self.rootPath)]
-
-        with open(os.path.join(self.project_dir.name, "project.yml"), "w+") as fd:
-            metadata: typing.Mapping[str, object] = {
-                "name": "Temporary project",
-                "description": "Temporary project",
-                "repo": yaml.safe_load(self.repos),
-                "modulepath": modulepath,
-                "downloadpath": "libs",
-                "install_mode": install_mode.value,
-            }
-            yaml.dump(metadata, fd)
-
-        v2_metadata_file: str = os.path.join(self.rootPath, module.ModuleV2.MODULE_FILE)
-        v1_metadata_file: str = os.path.join(self.rootPath, module.ModuleV1.MODULE_FILE)
-
-        module_name: Optional[str] = None
-        if os.path.exists(v2_metadata_file):
-            mv2 = module.ModuleV2(project=None, path=self.rootPath)
-            module_name = mv2.name
-        elif os.path.exists(v1_metadata_file):
-            mv1 = module.ModuleV1(project=None, path=self.rootPath)
-            module_name = mv1.name
-
-        if not module_name:
-            error_message: str = (
-                "The Inmanta extension only works on projects and modules. "
-                "Please make sure the current workspace is a valid project "
-                "(https://docs.inmanta.com/inmanta-service-orchestrator/latest/model_developers/project_creation.html) or "
-                "module (https://docs.inmanta.com/inmanta-service-orchestrator/latest/model_developers/module_creation.html)."
-            )
-            raise InvalidExtensionSetup(error_message)
-
-        with open(os.path.join(self.project_dir.name, "main.cf"), "w+") as fd:
-            fd.write(f"import {module_name}\n")
+    async def compile_and_anchor(self) -> None:
+        """
+        Perform a compile and compute an anchormap for the currently open folder.
+        """
 
-        return self.project_dir.name
+        def sync_compile_and_anchor_folder(folder: Folder):
+            logger.info(f"compile and anchor for folder {folder}")
 
-    async def compile_and_anchor(self) -> None:
-        def sync_compile_and_anchor() -> None:
-            def setup_project():
+            def setup_project(folder: Folder):
                 # Check that we are working inside an existing project:
-                project_file: str = os.path.join(self.rootPath, module.Project.PROJECT_FILE)
-                if os.path.exists(project_file):
-                    project_dir: str = self.rootPath
-
+                if folder.inmanta_project_dir:
+                    project_dir: str = folder.get_project_dir()
+                    logger.info(f"using  existing project {project_dir}")
                 else:
-                    # Create a project in the vscode temp folder
-                    project_dir = self.create_tmp_project()
+                    # Create a temporary project
+                    project_dir = folder.create_tmp_project()
```

Comment:
memory leak? This creates a new project for each compile but it is never cleaned up. Either the method should be idempotent (and its docstring should mention it) or we should clean this up after each compile.

Reply:
Unless I'm missing an edge case there shouldn't be a memory leak. The flow is as following:

- During initialization the InmantaLSHandler [registers](https://github.com/inmanta/vscode-inmanta/blob/5e1fd05c03b10a4dc82915ab3591ffebf09bc80b/server/src/inmantals/server.py#L260) the active folder  (It used to register all folders, but it will be changed to only register the active folder)
- First compile: create a temporary project
- Following compiles: use the already created temporary project
- During shutdown: [cleanup](https://github.com/inmanta/vscode-inmanta/blob/5e1fd05c03b10a4dc82915ab3591ffebf09bc80b/server/src/inmantals/server.py#L451) all temporary projects (same, this used to cleanup all registered folders, but there will be only the active one soon)

Reply:
Ah I see, I overlooked the third bullet point. I withdraw this comment.